Add Gap Mesh Generator

miaoyinb opened the pull request:
closes #32119

GiudGiud commented on framework/include/utils/MooseMeshUtils.h:
```diff
@@ -470,4 +470,41 @@ void copyIntoMesh(MeshGenerator & mg,
+                       const bool loop,
+                       const BoundaryName & start_boundary,
+                       const BoundaryName & end_boundary,
+                       const Real & max_elem_size);
```
```suggestion
                       const Real max_elem_size);
```

GiudGiud commented on modules/reactor/doc/content/source/meshgenerators/GapMeshGenerator.md:
```diff
@@ -0,0 +1,40 @@
+
+## Overview
+
+In the mesh generation for reactor geometries, it is often necessary to create gaps that are conformal to the external boundaries of components. The gap is meant to have a uniform thickness throughout the geometry. This `GapMeshGenerator` class provides functionality to generate a polyline mesh (similar to the meshes generated by [`PolyLineMeshGenerator`](/PolyLineMeshGenerator.md)) with a specif ...
```
```suggestion
In the mesh generation for reactor geometries, it is often necessary to create unmeshed gaps along the external boundaries of components. The gap is meant to have a uniform thickness all around the boundary. This `GapMeshGenerator` class provides functionality to generate a polyline mesh (similar to the meshes generated by [`PolyLineMeshGenerator`](/PolyLineMeshGenerator.md)) with a specified gap thickness based on an 2D input mesh residing in the XY plane. The input mesh must be a connected mesh with only one outer boundary manifold. The generated polyline mesh can then be used for further mesh generation processes, such as [`XYDelaunayGenerator`](/XYDelaunayGenerator.md).
```

GiudGiud commented on modules/reactor/doc/content/source/meshgenerators/GapMeshGenerator.md:
```diff
@@ -0,0 +1,40 @@
+!media meshgenerators/refined_gap.png
+      style=display: block;margin-left:auto;margin-right:auto;width:40%;
+      id=refined_gap
+      caption=A refined polyline mesh that is generated based on a set of patterned hexagons with a gap.
```
```suggestion
      caption=A refined polyline mesh that is generated around a set of patterned hexagons with a gap.
```

GiudGiud commented on modules/reactor/include/meshgenerators/GapMeshGenerator.h:
```diff
@@ -0,0 +1,55 @@
+ * Generate a polyline mesh that is based on an input 2D-XY mesh. The 2D-XY mesh needs to be a
+ * connected mesh with only one outer boundary manifold. The polyline mesh generated along with the
+ * boundary of the input mesh form a gap with a specified thickness. The mesh can further be used by
+ * XYDelaunayGenerator to generate a triangulation mesh that takes the gap into account.
```
```suggestion
 * boundary of the input mesh form a gap with a specified thickness.
```

GiudGiud commented on modules/reactor/include/meshgenerators/GapMeshGenerator.h:
```diff
@@ -0,0 +1,55 @@
+   * @param p3 Third point
+   * @return true if the three points are collinear, false otherwise
+   */
+  bool isPointsColinear(const Point & p1, const Point & p2, const Point & p3) const;
```
this should go in a utils if it s not there already

GiudGiud commented on modules/reactor/src/meshgenerators/GapMeshGenerator.C:
```diff
@@ -0,0 +1,219 @@
+#include "libmesh/parsed_function.h"
+#include "libmesh/poly2tri_triangulator.h"
+#include "libmesh/unstructured_mesh.h"
+#include "DelimitedFileReader.h"
```
```suggestion
#include "MooseUtils.h"
#include "DelimitedFileReader.h"

#include "libmesh/elem.h"
#include "libmesh/enum_to_string.h"
#include "libmesh/int_range.h"
#include "libmesh/mesh_modification.h"
#include "libmesh/mesh_serializer.h"
#include "libmesh/mesh_triangle_holes.h"
#include "libmesh/parsed_function.h"
#include "libmesh/poly2tri_triangulator.h"
#include "libmesh/unstructured_mesh.h"
```

GiudGiud commented on modules/reactor/src/meshgenerators/GapMeshGenerator.C:
```diff
@@ -0,0 +1,219 @@
+#include "libmesh/poly2tri_triangulator.h"
+#include "libmesh/unstructured_mesh.h"
+#include "DelimitedFileReader.h"
+
```
```suggestion
```

GiudGiud commented on modules/reactor/src/meshgenerators/GapMeshGenerator.C:
```diff
@@ -0,0 +1,219 @@
+  }
+
+  // To ensure no overlapping, we need to check set of four points
+  // p1 and p2 should be the pair of points before and after shifting
```
```suggestion
  // p1 and p2 should be the pair of points before shifting the side
```
?

GiudGiud commented on modules/reactor/src/meshgenerators/GapMeshGenerator.C:
```diff
@@ -0,0 +1,219 @@
+                "Each vertex should be connected to exactly two sides in a polygon.");
+
+    const Point original_pt = *(ply_mesh_u->node_ptr(node_id));
+    const Point move_dir = (normal_vecs.front() + normal_vecs.back()).unit();
```
```suggestion
    // Form an average normal at the vertex from the two connected sides' normals
    const Point move_dir = (normal_vecs.front() + normal_vecs.back()).unit();
```

GiudGiud commented on modules/reactor/src/meshgenerators/GapMeshGenerator.C:
```diff
@@ -0,0 +1,219 @@
+        _thickness /
+        std::sqrt((1.0 + (normal_vecs.front() * normal_vecs.back()) /
+                             (normal_vecs.front().norm() * normal_vecs.back().norm())) /
+                  2.0);
```
can you add a comment on the scaling factor here? not sure I understand it

GiudGiud reviewed (changes requested):
GapMeshGenerator could be meshing the gap instead of circling it on one side.
Maybe GapLineMeshGenerator ? Or BoundaryOffsetLineGenerator ? WallOffsetLineGenerator?

Does this generator work to mesh a gap inside rather than outside? Like inside the clad rather than outside the fuel?

miaoyinb committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

miaoyinb committed:
rename MG tools to utils and more comments

miaoyinb committed:
move to geom utils

miaoyinb committed:
inward option

miaoyinb committed:
Add Gap Mesh Generator

closes #32119

miaoyinb committed:
large media update

GiudGiud commented on framework/src/utils/GeometryUtils.C:
```diff
@@ -277,4 +277,45 @@ boxCorners(const libMesh::BoundingBox & box, const Real factor)
+
+  const Real denom = a1 * b2 - a2 * b1;
+  // We should not worry about the parallel case here
+  // If there is an overlap issue, it will be captured by other line segments
```
```suggestion
```

GiudGiud commented on framework/include/utils/GeometryUtils.h:
```diff
@@ -167,4 +167,23 @@ bool pointOnEdge(const libMesh::Point & point,
+ * @param p3 First point of second line segment
+ * @param p4 Second point of second line segment
+ * @return true if the two line segments intersect, false otherwise
+ */
```
```suggestion
 * Parallel overlapping segments are considered not intersecting.
 */
```

miaoyinb replied:
Will leave it to the caller to decide what value to return for the parallel case.

GiudGiud commented on modules/reactor/doc/content/source/meshgenerators/GapLineMeshGenerator.md:
```diff
@@ -0,0 +1,42 @@
+## Overview
+
+In the mesh generation for reactor geometries, it is often necessary to create unmeshed gaps along the external boundaries of components. The gap is meant to have a uniform thickness all around the boundary. This `GapLineMeshGenerator` class provides functionality to generate a polyline mesh (similar to the meshes generated by [`PolyLineMeshGenerator`](/PolyLineMeshGenerator.md)) with a specified ...
+
```
I think there needs to be a caveat on the list of boundary ids

can you do non-contiguous groups of boundary IDs? Does it detect it when the boundaries are contiguous?

GiudGiud replied:
nevermind you have it in the description of the class already.

GiudGiud commented on modules/reactor/doc/content/source/meshgenerators/GapLineMeshGenerator.md:
```diff
@@ -0,0 +1,42 @@
+
+This mesh generator first utilizes the `libMesh` utility used by [`XYDelaunayGenerator`](/XYDelaunayGenerator.md) to extract the external boundary or the specified boundary of the input mesh ([!param](/Mesh/GapLineMeshGenerator/input)). The boundary vertices are then shifted outward along the averaged normal directions of the connected edges to form the designated gap thickness ([!param](/Mesh/Ga ...
+
+The gap can be created either outward or inward with respect to the boundary of the input mesh by specifying the gap direction ([!param](/Mesh/GapLineMeshGenerator/gap_direction)).
```
```suggestion
The gap can be created either outward or inward with respect to the boundary of the input mesh by specifying the gap direction ([!param](/Mesh/GapLineMeshGenerator/gap_direction)). This direction is relative to the orientation of the sideset used.
```

GiudGiud replied:
```suggestion
The gap can be created either outward or inward with respect to the boundary of the input mesh by specifying the gap direction ([!param](/Mesh/GapLineMeshGenerator/gap_direction)). This direction is defined with regards to the elements' side outwards normal, not the orientation of the sideset (if the boundary is a sideset).
```

GiudGiud commented on modules/reactor/doc/content/source/meshgenerators/GapLineMeshGenerator.md:
```diff
@@ -0,0 +1,42 @@
+
+## Examples
+
+A simple example shown here is based on a set of patterned hexagons. As shown in [simple_gap], the generated polyline mesh without refinement only contains one `EDGE2` element for each side of the hexagons with a gap from the original hexagons.
```
```suggestion
A simple example shown here is based on a set of patterned hexagons. As shown in [simple_gap], the generated polyline mesh without refinement only contains one `EDGE2` element for each side of the hexagons with a gap from the original hexagons' sides.
```

GiudGiud commented on framework/src/utils/GeometryUtils.C:
```diff
@@ -277,4 +277,45 @@ boxCorners(const libMesh::BoundingBox & box, const Real factor)
 }
 
+bool
+isPointsColinear(const Point & p1, const Point & p2, const Point & p3)
```
```suggestion
arePointsColinear(const Point & p1, const Point & p2, const Point & p3)
```

GiudGiud reviewed (commented):
looks great!
some nitpicky comments. next is broken rn anyway

miaoyinb committed:
Apply suggestions from code review
- doco precisions
- avoid recalculating counter
- more checks on segment intersections

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 13922cef7b6da71adc754ba37a6ba4952848d908
+Subproject commit 8c17379b235e519eaeb7948adac2df390a6bd3ea
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3493282/), step Clang format on e7e87a5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/32120/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/32120/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4545b0941638efd8d6a23e44fbd4a3aa7e2f58a8`

loganharbour commented:
Reminder to add sweeps and valgrind to these significant mesh generator tests.

This broke devel: https://civet.inl.gov/event/274645/
